File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/connection/url/MediaServerUrlProvider.java
package com.lasthopesoftware.bluewater.client.connection.url;

import com.lasthopesoftware.bluewater.shared.IoCommon;

import java.io.UnsupportedEncodingException;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLEncoder;

public class MediaServerUrlProvider implements IUrlProvider {

	private final URL baseUrl;
	private final String authCode;
	private final byte[] certificateFingerprint;

	public MediaServerUrlProvider(String authCode, String ipAddress, int port, byte[] certificateFingerprint) throws MalformedURLException {
		this(authCode, new URL(IoCommon.httpsUriScheme, ipAddress, port, ""), certificateFingerprint);
	}

	public MediaServerUrlProvider(String authCode, String ipAddress, int port) throws MalformedURLException {
		this(authCode, new URL(IoCommon.httpUriScheme, ipAddress, port, ""));
	}

	public MediaServerUrlProvider(String authCode, URL baseUrl) throws MalformedURLException {
		this(authCode, baseUrl, new byte[0]);
	}

	private MediaServerUrlProvider(String authCode, URL baseUrl, byte[] certificateFingerprint) throws MalformedURLException {
		this.authCode = authCode;
		this.baseUrl = new URL(baseUrl, "/MCWS/v1/");
		this.certificateFingerprint = certificateFingerprint;
	}

	public String getAuthCode() {
		return authCode;
	}

	@Override
	public byte[] getCertificateFingerprint() {
		return certificateFingerprint;
	}

	public String getBaseUrl() {
		return baseUrl.toString();
	}

	public String getUrl(String... params) {
		// Add base url
		if (params.length == 0) return getBaseUrl();

		final StringBuilder urlBuilder = new StringBuilder(getBaseUrl());

		// Add action
		urlBuilder.append(params[0]);

		// add arguments
		if (params.length > 1) {
			for (int i = 1; i < params.length; i++) {
				urlBuilder.append(i == 1 ? '?' : '&');

				final String param = params[i];

				final int equalityIndex = param.indexOf('=');
				if (equalityIndex < 0) {
					urlBuilder.append(encodeParameter(param));
					continue;
				}

				urlBuilder.append(encodeParameter(param.substring(0, equalityIndex))).append('=').append(encodeParameter(param.substring(equalityIndex + 1)));
			}
		}

		return urlBuilder.toString();
	}

	private static String encodeParameter(String parameter) {
		try {
			return URLEncoder.encode(parameter, "UTF-8").replace("+", "%20");
		} catch (UnsupportedEncodingException e) {
			return parameter;
		}
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/connection/url/MediaServerUrlProvider.kt
package com.lasthopesoftware.bluewater.client.connection.url

import com.lasthopesoftware.resources.uri.IoCommon
import java.io.UnsupportedEncodingException
import java.net.URL
import java.net.URLEncoder

class MediaServerUrlProvider constructor(
	override val authCode: String?,
	baseUrl: URL,
	override val certificateFingerprint: ByteArray
) : IUrlProvider {

	private val baseURL = URL(baseUrl, "/MCWS/v1/")

	constructor(authCode: String?, ipAddress: String?, port: Int)
		: this(authCode, URL(IoCommon.httpUriScheme, ipAddress, port, ""), ByteArray(0))

	constructor(authCode: String?, ipAddress: String?, port: Int, certificateFingerprint: ByteArray)
		: this(authCode, URL(IoCommon.httpsUriScheme, ipAddress, port, ""), certificateFingerprint)

	override val baseUrl: URL
		get() = baseURL

	override fun getUrl(vararg params: String): String { // Add base url
		val urlString = baseURL.toString()
		if (params.isEmpty()) return urlString

		val urlBuilder = StringBuilder(urlString)

		// Add action
		urlBuilder.append(params[0])

		if (params.size == 1) return urlBuilder.toString()

		// add arguments
		for (i in 1 until params.size) {
			urlBuilder.append(if (i == 1) '?' else '&')
			val param = params[i]
			val equalityIndex = param.indexOf('=')
			if (equalityIndex < 0) {
				urlBuilder.append(encodeParameter(param))
				continue
			}
			urlBuilder.append(encodeParameter(param.substring(0, equalityIndex))).append('=').append(encodeParameter(param.substring(equalityIndex + 1)))
		}

		return urlBuilder.toString()
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (javaClass != other?.javaClass) return false

		other as MediaServerUrlProvider

		if (authCode != other.authCode) return false
		if (!certificateFingerprint.contentEquals(other.certificateFingerprint)) return false
		if (baseURL != other.baseURL) return false

		return true
	}

	override fun hashCode(): Int {
		var result = authCode?.hashCode() ?: 0
		result = 31 * result + certificateFingerprint.contentHashCode()
		result = 31 * result + baseURL.hashCode()
		return result
	}

	companion object {
		private fun encodeParameter(parameter: String): String {
			return try {
				URLEncoder.encode(parameter, "UTF-8").replace("+", "%20")
			} catch (e: UnsupportedEncodingException) {
				parameter
			}
		}
	}

}
